cli/internal/turbopath/absolute_path.go

```diff
@@ -0,0 +1,172 @@
+package turbopath
+
+import (
+	"io/ioutil"
+	"log"
+	"os"
+	"path/filepath"
+	"strings"
+)
+
+// dirPermissions are the default permission bits we apply to directories.
+const dirPermissions = os.ModeDir | 0775
+
+// ensureDir ensures that the directory of the given file has been created.
+func ensureDir(filename string) error {
+	dir := filepath.Dir(filename)
+	err := os.MkdirAll(dir, dirPermissions)
+	if err != nil && fileExists(dir) {
+		// It looks like this is a file and not a directory. Attempt to remove it; this can
+		// happen in some cases if you change a rule from outputting a file to a directory.
+		log.Printf("Attempting to remove file %s; a subdirectory is required", dir)
+		if err2 := os.Remove(dir); err2 == nil {
+			err = os.MkdirAll(dir, dirPermissions)
+		} else {
+			return err
+		}
+	}
+	return err
+}
+
+var nonRelativeSentinel string = ".." + string(filepath.Separator)
+
+// dirContainsPath returns true if the path 'target' is contained within 'dir'
+// Expects both paths to be absolute and does not verify that either path exists.
+func dirContainsPath(dir string, target string) (bool, error) {
+	// In Go, filepath.Rel can return a path that starts with "../" or equivalent.
+	// Checking filesystem-level contains can get extremely complicated
+	// (see https://github.com/golang/dep/blob/f13583b555deaa6742f141a9c1185af947720d60/internal/fs/fs.go#L33)
+	// As a compromise, rely on the stdlib to generate a relative path and then check
+	// if the first step is "../".
+	rel, err := filepath.Rel(dir, target)
+	if err != nil {
+		return false, err
+	}
+	return !strings.HasPrefix(rel, nonRelativeSentinel), nil
+}
+
+// fileExists returns true if the given path exists and is a file.
+func fileExists(filename string) bool {
+	info, err := os.Lstat(filename)
+	return err == nil && !info.IsDir()
+}
```

Comment:
These are duplicated utils from `fs/fs.go` (with the caveat of being unexported). They will be cleaned up in a followup PR.